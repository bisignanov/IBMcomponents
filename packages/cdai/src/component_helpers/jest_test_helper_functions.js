//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React from 'react';
import { shallow, mount } from 'enzyme';

export const getJSXForComponent = (
  component,
  defaultProps = {},
  customProps = {}
) => {
  return React.createElement(component, { ...defaultProps, ...customProps });
};

export const mountComponent = (jsx, returnShallow = true) => {
  let component, componentInstance;

  if (returnShallow) {
    component = shallow(jsx);
  } else {
    component = mount(jsx);
  }

  componentInstance = component.instance();

  let unmount = () => {
    unmountComponent(component);
  };

  return {
    component,
    componentInstance,
    unmount,
  };
};

export const unmountComponent = (component) => {
  component && component.length > 0 && component.unmount();
};
