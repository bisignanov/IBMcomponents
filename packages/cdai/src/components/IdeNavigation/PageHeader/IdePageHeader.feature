Feature: IDE Page Header component
  As an ICP4I product
  I want to be able to embed a common page header component.

Background:
    Given Setup of the IdePageHeader world

Scenario Outline: When I provide a set properties, these are rendered by the IdePageHeader component
    Given a set of properties called "<props>" provided to the component
    When the component is rendered
    Then the content of "<props>" is rendered as expected

    Examples:
    | props                                 |
    | none                                  |
    | breadcrumb                            |
    | tabs                                  |
    | breadcrumb,heading                    |
    | breadcrumb,tabs                       |
    | breadcrumb,others                     |
    | breadcrumb,custom                     |
    | breadcrumb,heading,tabs               |
    | breadcrumb,heading,others             |
    | breadcrumb,heading,custom             |
    | breadcrumb,heading,tabs,others        |
    | breadcrumb,heading,tabs,custom        |
    | breadcrumb,heading,tabs,others,custom |
    | header,headerRenderFunction           |
    | headerRenderFunction                  |
  