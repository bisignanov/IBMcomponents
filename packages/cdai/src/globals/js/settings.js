const settings = {
  idePrefix: 'ide',
};
module.exports = settings;
