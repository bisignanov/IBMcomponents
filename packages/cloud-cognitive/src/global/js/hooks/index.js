export { useWindowScroll } from './useWindowScroll';
export { useWindowResize } from './useWindowResize';
export { useClickOutside } from './useClickOutside';
export { useCreateComponentFocus } from './useCreateComponentFocus';
export { useCreateComponentStepChange } from './useCreateComponentStepChange';
export { usePreviousValue } from './usePreviousValue';
export { useResetCreateComponent } from './useResetCreateComponent';
export { useValidCreateStepCount } from './useValidCreateStepCount';
