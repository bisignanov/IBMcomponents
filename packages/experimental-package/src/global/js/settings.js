import featureFlags from '../../../../feature-flags/feature-flags';

const expPrefix = 'exp';

export { expPrefix };

export { featureFlags };
