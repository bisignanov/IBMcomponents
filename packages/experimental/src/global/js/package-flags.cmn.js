//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaultFlags = {
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    TearSheet: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const enableAll = (section) => {
  const result = {};
  for (const key in section) {
    result[key] = true;
  }
  return result;
};

const checkEnabled = (flags, overrides = {}) => {
  const result = {};
  for (const key in flags) {
    if (overrides[key]?.enableAll) {
      // enable all for this section
      delete overrides[key].enableAll;
      result[key] = enableAll(flags[key]);
    } else {
      result[key] = { ...flags[key], ...(overrides[key] || {}) };
    }
  }
  return result;
};

const packageFlags = () => {
  let initialized = false;
  let flags;

  // doNotUse parameter is intended for dev flags
  // doNotUse.initAgain causes flags to be refreshed which is handy in tests
  // an override section containing enableAll=true turns on everything in that section - handy for storybook
  return (overrides, doNotUse) => {
    const defaults = { ...defaultFlags };

    if (!initialized || doNotUse?.initAgain === true) {
      initialized = true;

      flags = checkEnabled(defaults, overrides);
    }

    return flags;
  };
};

const getPackageFlags = packageFlags();

module.exports = getPackageFlags;
