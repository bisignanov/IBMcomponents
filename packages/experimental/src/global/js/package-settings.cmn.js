//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const defaultFlags = {
  // by default only released components are set to true
  component: {
    AboutModal: false,
    ActionBar: false,
    APIKeyDownlaoder: false,
    APIKeyModal: false,
    BreadcrumbWithOverflow: false,
    Card: false,
    ContextHeader: false,
    EmptyState: false,
    ExampleComponent: false,
    ExportModal: false,
    ImportModal: false,
    ModifiedTabs: false,
    Notifications: false,
    PageActionItem: false,
    PageHeader: false,
    RemoveDeleteModal: false,
    SidePanel: false,
    TagSet: false,
    Tearsheet: false,
    TearsheetNarrow: false,
    TearsheetShell: false,
    WebTerminal: false,
  },

  // feature level flags
  feature: {
    'a-new-feature': false,
  },
};

const applyOverrides = (target, overrides = {}) => {
  for (const key in overrides) {
    if (key !== 'enableAll' && key !== 'disableAll') {
      // enable all is special handled one level up
      if (typeof target[key] === 'object') {
        if (typeof overrides[key] === 'object') {
          const disableAll = (key === 'component' || key === 'feature') && overrides[key].disableAll === true;
          const enableAll = (key === 'component' || key === 'feature') && overrides[key].enableAll === true;

          if (disableAll || enableAll) {
            // special case loop through all component or feature and enable
            for (const childKey in target[key]) {
              target[key][childKey] = enableAll;
            }
          }

          // process everything else working through branch
          applyOverrides(target[key], overrides[key]);
        } else {
          // reject replacing object with primitive
          console.warn(
            `Attempt to replace setting ${key} with primitive type rejected`
          );
        }
      } else {
        // both primitive types replace
        target[key] = overrides[key];
      }
    }
  }
};

const defaultPrefix = 'exp';

const settings = {
  prefix: defaultPrefix,
  flags: {
    component: { ...defaultFlags.component },
    feature: { ...defaultFlags.feature },
  },
  isComponentEnabled: (component, byDefault = false) => {
    const componentName = component?.name || component;
    const flags = byDefault ? defaultFlags : settings.flags;

    return flags.component[componentName];
  },
  isFeatureEnabled: (feature, byDefault = false) => {
    const flags = byDefault ? defaultFlags : settings.flags;
    return flags.feature[feature];
  },
  overrideSettings: (overrides, reset) => {
    if (reset) {
      settings.prefix = defaultPrefix;
      settings.flags = {
        component: { ...defaultFlags.component },
        feature: { ...defaultFlags.feature },
      };
    }
    applyOverrides(settings, overrides);
  },
  replaceWithCare: (replacement) => {
    console.warn(`

Presumably you know what you are doing as ou have replaced package settings:

------
${JSON.stringify(settings)}
------

with:

++++++
${JSON.stringify(replacement)}
++++++

`);

    for (const key in settings) {
      delete settings[key];
    }

    for (const key in replacement) {
      settings[key] = replacement[key];
    }
  },
};

module.exports = settings;
