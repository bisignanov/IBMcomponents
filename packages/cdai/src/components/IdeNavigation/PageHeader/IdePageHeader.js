//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import React from 'react';
import PropTypes from 'prop-types';
import { settings } from 'carbon-components';
import { idAttribute } from '../../../component_helpers/IDHelper';
import IdePageContent from '../PageContent/IdePageContent.js';
import { idePrefix } from '../../../globals/js/settings';

const { prefix } = settings;
const renderBreadcrumb = (breadcrumbContent) => (
  <div
    key={`IdePageHeader--breadcrumb-container`}
    className={`${idePrefix}-page-header--breadcrumb-container`}
    {...idAttribute('IdePageHeader--breadcrumb-container')}
  >
    {breadcrumbContent}
  </div>
);

const renderHeader = (header, headingRenderFunction) => (
  <div
    key={`IdePageHeader--heading--container`}
    className={`${idePrefix}-page-header--heading-container`}
    {...idAttribute('IdePageHeader--heading-container')}
  >
    {header
      ? headingRenderFunction(
          <p
            key={`IdePageHeader--heading`}
            className={`${idePrefix}-page-header--heading`}
            {...idAttribute('IdePageHeader--heading')}
          >
            {header}
          </p>
        )
      : headingRenderFunction(null)}
  </div>
);

const renderSubtitle = (subtitle) => (
  <div
    key={`IdePageHeader--subtitle--container`}
    className={`${idePrefix}-page-header--subtitle-container`}
    {...idAttribute('IdePageHeader--subtitle-container')}
  >
    {subtitle ? (
      <p
        key={`IdePageHeader--subtitle`}
        className={`${idePrefix}-page-header--subtitle`}
        {...idAttribute('IdePageHeader--subtitle')}
      >
        {subtitle}
      </p>
    ) : null}
  </div>
);
const renderTabs = (tabContent) =>
  tabContent ? (
    <div
      key={`IdePageHeader--tabs-container`}
      className={`${idePrefix}-page-header--tabs-container`}
      {...idAttribute('IdePageHeader--tabs-container')}
    >
      {tabContent}
    </div>
  ) : (
    <div
      className={`${idePrefix}-page-header--tabs-container__empty`}
      key={`IdePageHeader--tabs-container`}
      {...idAttribute('IdePageHeader--tabs-container')}
    />
  );

const IdePageHeader = ({
  breadcrumb,
  headingRenderFunction,
  heading,
  subtitle,
  tabs,
  customContent,
  className,
  ...others
}) => {
  // depending on the case, render some aspects of the navigation, depending on the properties provided
  let navContent = null;
  if (breadcrumb && !heading && !tabs) {
    navContent = renderBreadcrumb(breadcrumb);
  } else if (!breadcrumb && !heading && tabs) {
    navContent = renderTabs(tabs);
  } else {
    navContent = [
      renderBreadcrumb(breadcrumb),
      renderHeader(heading, headingRenderFunction),
      renderSubtitle(subtitle),
      renderTabs(tabs),
    ];
  }

  return (
    <section
      {...idAttribute('IdePageHeader')}
      className={`${idePrefix}-page-header ${className}`}
      {...others}
    >
      <IdePageContent
        bleedClassName={`${idePrefix}-page-header__backgroundColour`}
      >
        <div className={`${prefix}--row`}>
          <div className={`${prefix}--col`}>{navContent}</div>
          {customContent ? (
            <div
              className={`${prefix}--col ${idePrefix}-page-header--custom-content-container`}
              {...idAttribute('IdePageHeader--custom-content-container')}
            >
              {customContent}
            </div>
          ) : null}
        </div>
      </IdePageContent>
    </section>
  );
};

IdePageHeader.defaultProps = {
  breadcrumb: null,
  headingRenderFunction: (headerJSX) => headerJSX,
  heading: null,
  subtitle: null,
  tabs: null,
  customContent: null,
  className: '',
};

IdePageHeader.propTypes = {
  /** optional - an instance of the Carbon Breadcrumb component representing you current navigation state */
  breadcrumb: PropTypes.node,
  /** optional - className. Use this to add any custom css you may need */
  className: PropTypes.string,
  /** optional - an area of bespoke JSX rendering. Use this if you want to position specific elements alongside/in this navigation element */
  customContent: PropTypes.node,
  /** optional - a string to display as the header text */
  heading: PropTypes.string,
  /** optional - a render function, which will pass as parameter one the JSX for your header. Use this to render other content alongisde the header (eg an icon)  */
  headingRenderFunction: PropTypes.func,
  /** optional - a string to display as the subtitle under the header text */
  subtitle: PropTypes.string,
  /** optional - an instance of the Carbon Tabs component, with child tabs contained within. Itended content of the Tab are react-router links to navigate you around the UI */
  tabs: PropTypes.node,
};

export default IdePageHeader;
