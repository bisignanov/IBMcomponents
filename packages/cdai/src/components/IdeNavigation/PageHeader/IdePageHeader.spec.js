//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import IdePageHeader from './IdePageHeader';
import {
  pageHeaderCoreProps,
  withSubtitle,
  bespokeJSXProp,
} from './test_assets/testProps.js';
import * as jth from '../../../component_helpers/jest_test_helper_functions.js';

describe('IdePageHeader unit tests', () => {
  let component, unmount;

  const mountTestComponent = (defaultProps = {}, props = {}) => {
    let test = jth.mountComponent(
      jth.getJSXForComponent(IdePageHeader, defaultProps, props)
    );
    component = test.component;
    unmount = test.unmount;
  };

  afterEach(() => {
    // if unmount defined, call it to clear up any mounted component
    unmount && unmount();
  });

  it('renders the expected structure', () => {
    mountTestComponent({ ...pageHeaderCoreProps, ...bespokeJSXProp });
    expect(component.getElements()).toMatchSnapshot();
  });

  it('renders the subtitle structure', () => {
    mountTestComponent({ ...pageHeaderCoreProps, ...withSubtitle });
    expect(component.getElements()).toMatchSnapshot();
  });
});
