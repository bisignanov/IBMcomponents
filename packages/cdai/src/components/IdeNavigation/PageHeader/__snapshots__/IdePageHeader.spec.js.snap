// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdePageHeader unit tests renders the expected structure 1`] = `
[
  <section
    className="ide-page-header storybook--use-width storybook--ide-page-header-heading"
    data-ui-id="IdePageHeader"
  >
    <IdePageContent
      bleedClassName="ide-page-header__backgroundColour"
      className=""
    >
      <div
        className="bx--row"
      >
        <div
          className="bx--col"
        >
          <div
            className="ide-page-header--breadcrumb-container"
            data-ui-id="IdePageHeader--breadcrumb-container"
          >
            <Breadcrumb
              data-ui-id="breadcrumb"
              noTrailingSlash={true}
            >
              <BreadcrumbItem>
                <a
                  href="#top"
                >
                  items
                </a>
              </BreadcrumbItem>
              <BreadcrumbItem
                isCurrentPage={true}
              >
                <div>
                  item
                </div>
              </BreadcrumbItem>
            </Breadcrumb>
          </div>
          <div
            className="ide-page-header--heading-container"
            data-ui-id="IdePageHeader--heading-container"
          >
            <p
              className="ide-page-header--heading"
              data-ui-id="IdePageHeader--heading"
            >
              my heading
            </p>
          </div>
          <div
            className="ide-page-header--subtitle-container"
            data-ui-id="IdePageHeader--subtitle-container"
          />
          <div
            className="ide-page-header--tabs-container"
            data-ui-id="IdePageHeader--tabs-container"
          >
            <Tabs
              scrollDebounceWait={150}
              scrollIntoView={true}
              selected={0}
              selectionMode="automatic"
              type="default"
            >
              <Tab
                data-ui-id="tab1"
                label="Tab one"
                onClick={[Function]}
                onKeyDown={[Function]}
                selected={false}
              />
              <Tab
                data-ui-id="tab2"
                label="Tab two"
                onClick={[Function]}
                onKeyDown={[Function]}
                selected={false}
              />
            </Tabs>
          </div>
        </div>
        <div
          className="bx--col ide-page-header--custom-content-container"
          data-ui-id="IdePageHeader--custom-content-container"
        >
          <div
            className="storybook--idepageheader-button"
            data-ui-id="custom"
          >
            <Button>
              Add new topic
            </Button>
          </div>
        </div>
      </div>
    </IdePageContent>
  </section>,
]
`;

exports[`IdePageHeader unit tests renders the subtitle structure 1`] = `
[
  <section
    className="ide-page-header storybook--use-width storybook--ide-page-header-heading"
    data-ui-id="IdePageHeader"
  >
    <IdePageContent
      bleedClassName="ide-page-header__backgroundColour"
      className=""
    >
      <div
        className="bx--row"
      >
        <div
          className="bx--col"
        >
          <div
            className="ide-page-header--breadcrumb-container"
            data-ui-id="IdePageHeader--breadcrumb-container"
          >
            <Breadcrumb
              data-ui-id="breadcrumb"
              noTrailingSlash={true}
            >
              <BreadcrumbItem>
                <a
                  href="#top"
                >
                  items
                </a>
              </BreadcrumbItem>
              <BreadcrumbItem
                isCurrentPage={true}
              >
                <div>
                  item
                </div>
              </BreadcrumbItem>
            </Breadcrumb>
          </div>
          <div
            className="ide-page-header--heading-container"
            data-ui-id="IdePageHeader--heading-container"
          >
            <p
              className="ide-page-header--heading"
              data-ui-id="IdePageHeader--heading"
            >
              my heading
            </p>
          </div>
          <div
            className="ide-page-header--subtitle-container"
            data-ui-id="IdePageHeader--subtitle-container"
          >
            <p
              className="ide-page-header--subtitle"
              data-ui-id="IdePageHeader--subtitle"
            >
              my subtitle
            </p>
          </div>
          <div
            className="ide-page-header--tabs-container"
            data-ui-id="IdePageHeader--tabs-container"
          >
            <Tabs
              scrollDebounceWait={150}
              scrollIntoView={true}
              selected={0}
              selectionMode="automatic"
              type="default"
            >
              <Tab
                data-ui-id="tab1"
                label="Tab one"
                onClick={[Function]}
                onKeyDown={[Function]}
                selected={false}
              />
              <Tab
                data-ui-id="tab2"
                label="Tab two"
                onClick={[Function]}
                onKeyDown={[Function]}
                selected={false}
              />
            </Tabs>
          </div>
        </div>
      </div>
    </IdePageContent>
  </section>,
]
`;
